lib/Containers/AtomicList.h

```diff
@@ -0,0 +1,279 @@
+//   _memoryThreshold * _maxHistory.
+// It is possible to take a snapshot of the current list and of all
+// historic lists. These snapshots may be traversed, but it is not allowed
+// to delete the Node pointers.
+// The type T must be an object which has a move constructor and has
+// a method called `memoryUsage` which estimates the memory usage
+// (including all substructures) in bytes. It should always return a
+// positive value, but this is intentionally not enforced.
+template<typename T>
+class BoundedList {
+  static_assert(std::is_object_v<T>, "T must be an object type");
+  static_assert(
+      std::is_same_v<decltype(std::declval<T>().memoryUsage()), size_t>,
+      "T must have a memoryUsage() method returning size_t");
+  std::atomic<std::shared_ptr<AtomicList<T>>> _current;
+  std::atomic<std::size_t> _memoryUsage;
+  std::vector<std::shared_ptr<AtomicList<T>>> _history;
+  std::vector<std::shared_ptr<AtomicList<T>>> _trash;
+  size_t _ringBufferPos;
+  mutable std::mutex
+      _mutex;  // protecting write access to _current and _history
+  const std::size_t _memoryThreshold;
+  const std::size_t _maxHistory;
+
+ public:
+  // The actual memory usage is max_history * memory_threshold and some minor
+  // overshooting is possible!
+  BoundedList(std::size_t memoryThreshold, std::size_t maxHistory)
+      : _current(std::make_shared<AtomicList<T>>()),
+        _memoryUsage(0),
+        _history(maxHistory),
+        _ringBufferPos(0),
+        _memoryThreshold(memoryThreshold),
+        _maxHistory(maxHistory) {
+    // The constructor exception could be more specific:
+    if (memoryThreshold == 0 || maxHistory < 2) {
+      throw std::invalid_argument(
+          "BoundedList: memoryThreshold must be > 0 and maxHistory must be >= "
+          "2");
+    }
+  }
+
+  void prepend(T&& value) {
+    // Can throw in out of memory situations!
+    auto mem_usage = value.memoryUsage();
+    // This load may synchronize with a store which has inserted a new
+    // list in this very method below, therefore acquire semantics.
+    std::shared_ptr<AtomicList<T>> current =
+        _current.load(std::memory_order_acquire);
+    current->prepend(std::move(value));
+
+    // We assume throughout that `size_t` is at least 64bits and over- or
+    // underflow is not a problem.
+    size_t prev_usage =
+        _memoryUsage.fetch_add(mem_usage, std::memory_order_relaxed);
+    size_t newUsage = prev_usage + mem_usage;
+    if (prev_usage < _memoryThreshold && newUsage >= _memoryThreshold) {
+      // We are the chosen one which is supposed to rotate _current into
+      // _history!
+      std::lock_guard<std::mutex> guard(_mutex);
+      if (_current.load(std::memory_order_relaxed) == current) {
+        // If somebody else came before us, we do not bother to rotate!
+        _current.store(std::make_shared<AtomicList<T>>(),
+                       std::memory_order_seq_cst);
```

Comment:
Out of interest: Why does it not suffice to use `std::memory_order_release` here when we synchronize with the acquire above?

Reply:
I thought to use the "default" since it is not time critical here. But as I see, this is more confusing than using the minimal needed memory order. Changed.